fq.go

```diff
@@ -0,0 +1,80 @@
+package fq
+
+import (
+	"math"
+	"sync"
+)
+
+type fqscheduler struct {
+	lock   *sync.Mutex
+	queues []*Queue
+	vt     *virtimer
+}
+
+func (q *fqscheduler) chooseQueue(packet *Packet) *Queue {
+	for _, queue := range q.queues {
+		if packet.key == queue.key {
+			return queue
+		}
+	}
+	// fmt.Printf("packet w/ key: %v\n", packet.key)
+	panic("no matching queue for packet")
+}
+
+func newfqscheduler(queues []*Queue) *fqscheduler {
+	fq := &fqscheduler{
+		lock:   &sync.Mutex{},
+		queues: queues,
+		vt:     &virtimer{},
+	}
+	return fq
+}
+
+func (q *fqscheduler) enqueue(packet *Packet) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	queue := q.chooseQueue(packet)
+	queue.enqueue(packet)
+	q.updateTime(packet, queue)
+
+}
+
+type virtimer struct {
+	round uint64
+}
+
+func (vt *virtimer) now() uint64 {
+	vt.round++
```

Comment:
This is wrong.  Reading the virtual clock should not advance it.  We do _not_ have the coincidence that the virtual clock is read exactly once per round.  In this commit, the virtual clock is read whenever a packet _arrives_ --- which is _not_ what should advance the virtual clock.

Reply:
Yes sorry, this is my mistake.  I believe that I had overlooked this when I was syncing updates from one branch to another. I will update this now.